api/src/main/scala/de/zalando/play/controllers/SecurityExtractors.scala

```diff
@@ -0,0 +1,129 @@
+
+  def queryApiKey[User >: Any]: String => RequestHeader => (String => User) => Future[Option[User]] =
+    name => header => convertToUser => header.queryString.get(name).flatMap(_.headOption) map convertToUser
+
+  def oAuthPassword[User >: Any]: Seq[String] => String => RequestHeader => (JsValue => User) => Future[Option[User]] =
+    scopes => tokenUrl => header => convertToUser => {
+    val futureResult = header.headers.get("Authorization").flatMap(decodeBearer).map { token: String =>
+      checkOAuthToken(tokenUrl, token, scopes:_*)
+    } match {
+      case None => Future.successful(None)
+      case Some(f) => f
+    }
+    futureResult.map(_.map(convertToUser))
+  }
+
+  def oAuth[User >: Any]: (String*) => RequestHeader => (Seq[String] => User) => Future[Option[User]] =
+    scopes => header => convertUer => ???
+}
+
+trait BasicAuthSecurityExtractor {
+  private val basicSt = "basic "
+
+  protected def decodeBasicAuth(auth: String): Option[(String, String)] = {
+    lazy val basicReqSt = auth.substring(0, basicSt.length())
+    lazy val basicAuthSt = auth.replaceFirst(basicReqSt, "")
+    lazy val decoder = new BASE64Decoder() //BESE64Decoder is not thread safe
+    lazy val decodedAuthSt = new String(decoder.decodeBuffer(basicAuthSt), "UTF-8")
+    lazy val usernamePassword = decodedAuthSt.split(":")
+
+    if (auth.length() < basicSt.length()) None
+    else if (basicReqSt.toLowerCase() != basicSt) None
+    else if (usernamePassword.length != 2) None
+    else Some(usernamePassword.head, usernamePassword.last)
+  }
+}
+
+trait OAuthCommon {
+  private val bearer = "Bearer "
+
+  protected def decodeBearer(auth: String): Option[String] =
+    if (auth.length < bearer.length) None
+    else if (auth.substring(0, bearer.length).toLowerCase != bearer) None
+    else Some(auth.substring(bearer.length))
+}
+/**
+  * This implementation implements token verification in following way:
+  *
+  * 1. If there is a ':token' part in the token url, it will be replaced by the url encoded token value
+  *   then, the GET request with the resulting url is issued and response is parsed as JSON,
+  *   as defined in https://tools.ietf.org/html/rfc7662
+  *
+  * 2. If no ':token' part present in the url, the POST request with Content-Type: application/x-www-form-urlencoded
+  *   and token parameter performed and the result parsed as JSON. This behavior is as
+  *   defined in https://tools.ietf.org/html/rfc7662
+  */
+trait OAuthResourceOwnerPasswordCredentialsFlow extends OAuthCommon {
+  import play.api.libs.ws._
+  import play.api.Play.current
+  private val placeHolder = ":token"
+  protected def checkOAuthToken(tokenUrl: String, token: String, requiredScopes: String*): Future[Option[JsValue]] = {
+    val escapedToken = URLEncoder.encode(token, "UTF-8")
+    val (method, url, body) =
+      if (tokenUrl.contains(placeHolder)) ("GET", tokenUrl.replaceAllLiterally(placeHolder, escapedToken), "")
```

Comment:
@alexkops thanks for feedback!

I expect this code to change a lot as two complex oAuth2 flows still not implemented.